engine/src/main/java/org/terasology/engine/rendering/opengl/OpenGLMesh.java

```diff
@@ -93,188 +65,111 @@ public TFloatList getVertices() {
-        partSizes.add(VERTEX_SIZE);
+        this.disposalAction.dispose();
+        this.disposalAction.vao = GL30.glGenVertexArrays();
+        this.disposalAction.vbo = GL30.glGenBuffers();
+        this.disposalAction.ebo = GL30.glGenBuffers();
 
-        if (newData.getTexCoord0() != null && newData.getTexCoord0().size() / TEX_COORD_0_SIZE == vertexCount) {
-            parts.add(newData.getTexCoord0().iterator());
-            partSizes.add(TEX_COORD_0_SIZE);
-            texCoord0Offset = vertexSize * FLOAT_SIZE;
-            vertexSize += TEX_COORD_0_SIZE;
-            hasTexCoord0 = true;
-        }
-        if (newData.getTexCoord1() != null && newData.getTexCoord1().size() / TEX_COORD_1_SIZE == vertexCount) {
-            parts.add(newData.getTexCoord1().iterator());
-            partSizes.add(TEX_COORD_1_SIZE);
-            texCoord1Offset = vertexSize * FLOAT_SIZE;
-            vertexSize += TEX_COORD_1_SIZE;
-            hasTexCoord1 = true;
-        }
-        if (newData.getNormals() != null && newData.getNormals().size() / NORMAL_SIZE == vertexCount) {
-            parts.add(newData.getNormals().iterator());
-            partSizes.add(NORMAL_SIZE);
-            normalOffset = vertexSize * FLOAT_SIZE;
-            vertexSize += NORMAL_SIZE;
-            hasNormal = true;
-        }
-        if (newData.getColors() != null && newData.getColors().size() / COLOR_SIZE == vertexCount) {
-            parts.add(newData.getColors().iterator());
-            partSizes.add(COLOR_SIZE);
-            colorOffset = vertexSize * FLOAT_SIZE;
-            vertexSize += COLOR_SIZE;
-            hasColor = true;
-        }
-        stride = vertexSize * FLOAT_SIZE;
-        indexCount = newData.getIndices().size();
+        List<Layout> layouts = newData.getLayouts();
+        List<Layout> targets = new ArrayList<>();
 
-        createVertexBuffer(parts, partSizes, vertexCount, vertexSize);
-        createIndexBuffer(newData.getIndices());
-        getBound(newData, aabb);
-    }
+        int stride = 0;
+        for (Layout layout : layouts) {
+            if (layout.hasContent()) {
+                stride += layout.bytes();
+                targets.add(layout);
+            }
+        }
 
-    private void createVertexBuffer(List<TFloatIterator> parts, TIntList partSizes, int vertexCount, int vertexSize) {
-        FloatBuffer vertexBuffer = BufferUtils.createFloatBuffer(vertexSize * vertexCount);
-        for (int v = 0; v < vertexCount; ++v) {
-            for (int partIndex = 0; partIndex < parts.size(); ++partIndex) {
-                TFloatIterator part = parts.get(partIndex);
-                for (int i = 0; i < partSizes.get(partIndex); ++i) {
-                    vertexBuffer.put(part.next());
-                }
+        ByteBuffer buffer = BufferUtils.createByteBuffer(stride * newData.getSize());
+        for (int x = 0; x < newData.getSize(); x++) {
+            for (Layout layout : targets) {
+                layout.write(x, buffer);
```

Comment:
Shouldn't this be taking the stride into account too?

Reply:
I pack all the data local to each other

XYZRGB|XYZRGB|XYZRGB|XYZRGB

stride is the length of the payload so XYZRGB. the offset into the same position for a position in the chunk is the lenght of the payload.

I could of packed it like this.

XYZXYZXYZXYZ | RGBRGBRGBRGB. the stride would be zero and the offset would be some number into the assoicated sections.

Reply:
Oh, okay. I missed the bit about it just adding the data to the end of the buffer.